execution_engine/src/storage/trie/mod.rs

```diff
@@ -346,6 +350,96 @@ impl ::std::fmt::Debug for PointerBlock {
     }
 }
 
+/// Represents a Merkle Tree or a chunk of data with attached proof.
+/// Chunk with attached proof is used when the requested
+/// trie is larger than [ChunkWithProof::CHUNK_SIZE_BYTES].
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
+pub enum TrieOrChunkedData {
+    /// Represents a Merkle Trie.
+    Trie(Box<Trie<Key, StoredValue>>),
+    /// Represents a chunk of data with attached proof.
+    ChunkWithProof(ChunkWithProof),
+}
+
+impl Display for TrieOrChunkedData {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        match self {
+            TrieOrChunkedData::Trie(trie) => {
+                write!(f, "Trie({})", trie)
+            }
+            TrieOrChunkedData::ChunkWithProof(chunk) => {
+                write!(
+                    f,
+                    "ChunkWithProof {{ index = {}, hash = {} }}",
+                    chunk.proof().index(),
+                    chunk.proof().root_hash()
+                )
+            }
+        }
+    }
+}
+
+impl TrieOrChunkedData {
+    fn tag(&self) -> u8 {
+        match self {
+            TrieOrChunkedData::Trie(_) => 0,
+            TrieOrChunkedData::ChunkWithProof(_) => 1,
+        }
+    }
+}
+
+impl ToBytes for TrieOrChunkedData {
```

Comment:
Can you also implement `ToBytes::write_bytes` manually and write data to the mutable vector that is passed in as an argument of the method?

Reply:
I think we will also need a `write_bytes` implementation for `ChunkWithProof`.